Ood shiny admin

MorganRodgers opened the pull request:
These changes convert pseudofun into the Shiny Admin Tool.

Commit:
Initial commit of BSIR Admin Tool

The BSIR Admin Tool allows CRUDing a mapping between Users, Applications and Datasets. It publishes that mapping in both an Sqlite3 database as well as a YAML document. Callbacks are installed on the create and destroy handlers for Mapping, and will overwrite the YAML storage with the new state. The YAML storage is never read by the Admin app.

On the list view mappings that have become invalid (missing directories or bad permissions) since creation are called out with a FontAwesome warning icon.

Most of the pseudofun specific code has been removed, but there are bound to be some remanents. The new mapping form is currently a straight post request, so it is not sticky in case of failure. Likewise it would be good to add more information to the valid field in the list view (in either a modal or a tool tip), so that the admin knows exactly what remedy is required.

Commit:
Implement permissions check for Mapping::is_still_valid?

Permissions may change or files may be deleted outside of this app, ensure that we can detect when a breaking change has occurred.

Commit:
Remove references to BSIR

Commit:
Ran bundle update against gems with known vulnerabilities

Commit:
Upgraded to Rails 4.2.10 and fixed spelling error

Commit:
Added warning when attempting to create new mapping with no apps installed

Commit:
Reordered model creation in MappingController to validate parameters before FACL setting

Commit:
Removed CC license and updated README to remove references to pseudofun

ericfranz commented:
One initial observation: to follow the Rails conventions for resources, it would be appropriate to use the plural "mappings" for the routes, controller naming, and view directory while for the model object retain the singular "Mapping". Thus:

```
# config/routes.rb
resources :mappings

# app/controllers/mappings_controller.rb
class MappingsController
  def index
    @mappings = Mapping.all
  end

# ... etc
end

# app/views/mappings/ for controller views
```

This naming convention helps to differentiate from a singular resource. (see http://guides.rubyonrails.org/v4.2/action_controller_overview.html#controller-naming-convention and http://guides.rubyonrails.org/v4.2/routing.html#singular-resources)

ericfranz commented:
Following up on the previous comment: converting the controller/route/views directory names (and other names as appropriate) to the plural "mappings" will let you use the default Rails route helpers in the views and controllers.

For example, if I login to webdev02 and do

```
scl enable nodejs010 rh-ruby22 rh-ror42 -- bash
cd tmp
rails new shiny_admin --skip-bundle
cd shiny_admin
bin/rails g scaffold mapping
```

I end up with `app/models/mapping.rb` and `test/models/mapping_test.rb` but everything else uses the plural `mappings` and when I view the routes it shows the helpers:

```
efranz@webdev02:~/tmp/shiny_admin (master)$ bin/rake routes
      Prefix Verb   URI Pattern                  Controller#Action
    mappings GET    /mappings(.:format)          mappings#index
             POST   /mappings(.:format)          mappings#create
 new_mapping GET    /mappings/new(.:format)      mappings#new
edit_mapping GET    /mappings/:id/edit(.:format) mappings#edit
     mapping GET    /mappings/:id(.:format)      mappings#show
             PATCH  /mappings/:id(.:format)      mappings#update
             PUT    /mappings/:id(.:format)      mappings#update
             DELETE /mappings/:id(.:format)      mappings#destroy
```

You append to the prefix `_path` or `_url` depending on what you want to use. Typically you use `_path`. 

Thus instead of `redirect_to mapping_index_path` you would do `redirect_to mappings_path`. Though the singlular resource route fortunately has most of the same mappings as the plural, thus `<%= link_to 'Edit', edit_mapping_path(@mapping.id), class: 'btn btn-default' %>` remains the same.

ericfranz commented:
You have very good form with methods that require imperative coding style, explicitly separating variable initialization at the top, method body, and return value at the bottom, which makes for readable code.

ericfranz commented:
The end refactor goal for our Rails apps so far have been:

1. Keep controllers as close to the style produced by the original scaffold as possible. So for example, in the default mappings controller produced by scaffolding `MappingsController#create` is:

    ```ruby
    # POST /mappings
    # POST /mappings.json
    def create
      @mapping = Mapping.new(mapping_params)

      respond_to do |format|
        if @mapping.save
          format.html { redirect_to @mapping, notice: 'Mapping was successfully created.' }
          format.json { render :show, status: :created, location: @mapping }
        else
          format.html { render :new }
          format.json { render json: @mapping.errors, status: :unprocessable_entity }
        end
      end
    end
    ```
    But sometimes we need more complexity. Here is an example of attempting to maintain the simplicity of the style above in the controller while still accommodating more complex resource creation: https://github.com/OSC/ood-dashboard/blob/7c4fbf45c5cdd415a9d76a3fda5c05dc8daf172a/app/controllers/products_controller.rb#L44-L57. So this would be part of the rational for recommending to move more code into the model. A benefit of this besides making the controllers easy to read is that models typically are easier to test than controllers. In fact, Sandi Metz claimed when writing Rails apps she only writes tests for models and plain old ruby objects, not controllers.
2. Use a functional style with small methods whenever possible (i.e. https://robots.thoughtbot.com/sandi-metz-rules-for-developers for inspiration)
3. Avoid control flow if possible. And its corollary, avoid returning nil when possible.
4. Keep logic out of the views when possible.

We have mostly pulled this off in most of our Rails apps, though there are definite examples of violations. Any code suggestions I provide in the future in regard to a Rails app will probably be motivated by some of the above goals.

ericfranz commented on app/helpers/mapping_helper.rb:
```diff
@@ -0,0 +1,49 @@
+
+    # Sort the lists, keeping the two location types separate
+    installed_datasets.to_a.sort + non_std_location_datasets.to_a.sort
+  end
```
A functional approach to this would be to have a method `managed_datasets` or `installed_datasets` that returns the list of datasets in BISR's `dataset_root` and then do something like this:

```ruby
def known_datasets
  (installed_datasets | Mapping.select(:dataset).map(&:dataset).sort.to_a).uniq
end
```

The `|` preserves order so you get the result I think you were hoping for. Per our discussion of if there was a more appropriate place for this - you could put this particular method on the model itself in a block like:

```ruby
class Mapping

  class << self
    # class methods here
  end
```

Since the method is doing a db query better in the model than in a helper or view.

Also, while Set ensures uniqueness, in this case just using the `uniq` method on `Array` is probably good enough. I think using Set becomes beneficial if we would imagine manipulating the set after it was initialized.

ericfranz replied:
Since `installed_datasets` is sure to be unique, you could actually use Active Record's `distinct` which I think is an alias for `uniq`.

Might be more readable:

```ruby
def known_datasets
  installed_datasets | Mapping.select(:dataset).distinct.order(:dataset).map(&:dataset)
end
```

Of course in this case its best practice to use scopes (http://guides.rubyonrails.org/v4.2/active_record_querying.html#scopes) for these queries so something like:

```ruby
class Mapping
  scope :datasets, -> {  select(:dataset).distinct.order(:dataset) }
```

and then we simplify further:

```ruby
def known_datasets
  installed_datasets | Mapping.datasets.map(&:dataset)
end
```

Well maybe scope is inappropriate here, and rather just a class method to return that array of datasets, because this seems odd `Mapping.datasets.map(&:dataset)`.

```ruby
class Mapping
  def self.datasets
    select(:dataset).distinct.order(:dataset).map(&:dataset)
  end
```

then

```ruby
def known_datasets
  installed_datasets | Mapping.datasets
end
```

Anyways ideas.

MorganRodgers replied:
I like that `scopes` can be chained, but I find the name `scope` unintuitive. Next time I'm in a room with the ActiveRecord developers remind me to ask them why they chose that 😉.

For the dataset case I like using a static method, but instead of doing the `Enumerable.map` it looks like I should prefer `ActiveRecord.pluck`.

ericfranz commented:
On my earlier comment on “functional style” a clarification would be that individual methods of a class would follow a functional style when possible. Outside of that we have tried to write good object oriented code and in Ruby favoring classes were we call new to then execute methods on as opposed to starting with static methods when there is no class state. It makes testing and refactoring (among  other things) a little easier.

Commit:
Removed failing tests from pseudofun

Commit:
Ensure that tmp/ exists with .keep

Commit:
Remove more references to pseudofun

Commit:
Removed another pseudofun reference and fixed env variables

ericfranz commented:
One more thought, an alternative to modifying the facls in ActiveRecord callbacks would be to have a custom save method on the model, something like:

```ruby
def save_and_set_facls
  return false unless valid? # run validations before - I think this is how you do it

  # apply facls, save, etc.
  save(false) # since we already ran validations, false means we don't run them again
end
```

then in the controller call `save_and_set_facls` instead of `save`. I think we have done this before in other apps though I don't remember where.

Commit:
FACL handling changes and other minor changes

Remove :articles resource from learning Rails
Moved FACL setting from controller to model life cycle hooks
Added paren for Mapping.find
Added check to see if admin owns file before attempting to alter FACL

Commit:
Renamed Mapping -> Mappings

Commit:
Removed edit view

Commit:
Updates to URL generation and fix for FACL permission check

Removed show template

Commit:
Actually includes the FACL fix plus removing two unused controller actions edit and show

Commit:
Assorted code review fixes

Fix delete URL
Moved no apps installed into own template to avoid having logic in templates
Since only new and delete is allowed the CRUD template has been moved into the new template, and removed CRUD template partial

Commit:
Assorted code review fixes

Added method/url comments to controller actions to be closer in style with other OSC repos
Moved Mapping CRUD safety checks out of controller and into class definition
Was unsuccessful refactoring flash statements into the redirect_to calls
Added a class method to Mapping to generate a unique list of datasets
Removed redundent alert and notice messages from application layout

Commit:
Assorted code review fixes

Removed redundant conversion of Array to Array, moved sort
Renamed attempt_(destory|save) to more descriptive name

Commit:
Updated prod settings to use $USER and $HOME
